### PR TITLE
Removing cached shard progress, adding guardrails for duplicate shard responses.

### PR DESCRIPTION
**Pull request by joshua-kim**
*Issue #, if available:*

*Description of changes:*

Removing cached shardmap that's used for shard syncs, as this can lead to unexpected behavior in partial failures. Also adding guardrails for duplicate shard responses.

Example logging:
```
2021-05-03 12:41:25,342 [main] WARN  com.amazonaws.services.kinesis.clientlibrary.proxies.KinesisProxy - Found duplicate shards in response when sync'ing from Kinesis. Request ids - [NONE]. Response - [{ShardId: shard-1,}, {ShardId: shard-2,}, {ShardId: shard-3,}, {ShardId: shard-4,}, {ShardId: shard-1,}, {ShardId: shard-2,}, {ShardId: shard-3,}, {ShardId: shard-4,}]
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/test/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxyTest.java [hunk 1]
```diff
@@ -483,6 +435,32 @@ public void testGetShardCacheAgeEmptyForcesUpdate() {
...
+        ListShardsResult lastPage = new ListShardsResult().withShards(shards.subList(2, shards.size())).withNextToken(null);
+        when(mockClient.listShards(any())).thenReturn(firstPage).thenThrow(new RuntimeException("Failed!")) // First call
+                .thenReturn(firstPage).thenReturn(lastPage); // second call
+
+        try {
+            proxy.getShardList();
+        } catch (Exception e) {
+            // Do nothing
+        }
+        assertEquals(shards, proxy.getShardList());
```
Comment (isurues) [inline src/test/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxyTest.java:451]: I think it's better to validate the shard Ids and the order as well here.

File: src/test/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxyTest.java [hunk 2]
```diff
@@ -483,6 +435,32 @@ public void testGetShardCacheAgeEmptyForcesUpdate() {
...
+        proxy.setCachedShardMap(null);
+        ListShardsResult firstPage = new ListShardsResult().withShards(shards.subList(0, 2)).withNextToken(NEXT_TOKEN);
+        ListShardsResult lastPage = new ListShardsResult().withShards(shards.subList(2, shards.size())).withNextToken(null);
+        when(mockClient.listShards(any())).thenReturn(firstPage).thenThrow(new RuntimeException("Failed!")) // First call
+                .thenReturn(firstPage).thenReturn(lastPage); // second call
+
+        try {
+            proxy.getShardList();
+        } catch (Exception e) {
+            // Do nothing
```
Comment (isurues) [inline src/test/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxyTest.java:449]: As we must get an exception here, let's add a boolean to track that and add an assert below for that boolean.

File: src/test/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxyTest.java [hunk 3]
```diff
@@ -483,6 +435,32 @@ public void testGetShardCacheAgeEmptyForcesUpdate() {
...
+        }
+        assertEquals(shards, proxy.getShardList());
+    }
+
+    @Test
+    public void testDuplicateShardResponseDedupedGracefully() {
+        proxy.setCachedShardMap(null);
+        ListShardsResult pageOfShards = new ListShardsResult().withShards(shards).withNextToken(null);
```
Comment (isurues) [inline src/test/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxyTest.java:457]: Does the "shards" list include duplicates? I don't think we are testing the duplicate case here.

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxy.java [hunk 4]
```diff
@@ -478,16 +476,21 @@ private boolean cacheNeedsTimeUpdate() {
...
+                    final Shard lastShard = pageOfShards.get(pageOfShards.size() - 1);
+                    if (lastShardId == null || lastShardId.compareTo(lastShard.getShardId()) < 0) {
+                        lastShardId = lastShard.getShardId();
+                    }
                 }
             } while (response.getStreamDescription().isHasMoreShards());
         }
-        List<Shard> shards = shardIterationState.getShards();
-        this.cachedShardMap = shards.stream().collect(Collectors.toMap(Shard::getShardId, Function.identity()));
+        final List<Shard> dedupedShards = new ArrayList<>(new LinkedHashSet<>(shards));
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxy.java:489]: Agree. Will add some logging for this.

File: src/test/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxyTest.java (hunk 2 above)
Comment (joshua-kim) [reply src/test/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxyTest.java:449]: Added a check for this.

File: src/test/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxyTest.java (hunk 1 above)
Comment (joshua-kim) [reply src/test/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxyTest.java:451]: Discussed offline, this is already checked for.

File: src/test/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxyTest.java (hunk 3 above)
Comment (joshua-kim) [reply src/test/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxyTest.java:457]: Your'e right... duplicated the shard in next revision...

File: src/main/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxy.java [hunk 5]
```diff
@@ -487,6 +489,9 @@ private boolean cacheNeedsTimeUpdate() {
             } while (response.getStreamDescription().isHasMoreShards());
         }
         final List<Shard> dedupedShards = new ArrayList<>(new LinkedHashSet<>(shards));
+        if (dedupedShards.size() < shards.size()) {
+            LOG.warn("Found duplicate child shards in ListShards response. Request ids - " + requestIds);
```
[the remark this answers is not in the document]
Comment (joshua-kim) [reply src/main/java/com/amazonaws/services/kinesis/clientlibrary/proxies/KinesisProxy.java]: Agree, updated the logging.

**Review by isurues — APPROVED** [review]
Looks good to me.
